File: day2/go/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func isInvalid(n int) bool {
	s := strconv.Itoa(n)
	length := len(s)

	// Try each possible pattern length (must repeat at least twice)
	for patternLen := 1; patternLen <= length/2; patternLen++ {
		// Pattern length must divide total length evenly
		if length%patternLen != 0 {
			continue
		}

		pattern := s[:patternLen]
		repetitions := length / patternLen

		if strings.Repeat(pattern, repetitions) == s {
			return true
		}
	}

	return false
}

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}

	file, _ := os.Open(filename)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Scan()
	line := scanner.Text()

	line = strings.TrimSuffix(line, ",")

	ranges := strings.Split(line, ",")
	sum := 0

	for _, r := range ranges {
		parts := strings.Split(r, "-")
		start, _ := strconv.Atoi(parts[0])
		end, _ := strconv.Atoi(parts[1])

		for n := start; n <= end; n++ {
			if isInvalid(n) {
				sum = sum + n
			}
		}
	}

	fmt.Println(sum)
}

File: day2/haskell/part2.hs
import System.Environment (getArgs)

main :: IO ()
main = do
    args <- getArgs
    let filename = if null args then "input.txt" else head args
    content <- readFile filename
    let ranges = map parseRange (takeWhile (/= "") (lines content))
        total = sum [n | (start, end) <- ranges, n <- [start..end], isInvalid n]
    print total

parseRange :: String -> (Int, Int)
parseRange line = (read start, read end)
  where
    (start, rest) = span (/= '-') line
    end = tail rest

isInvalid :: Int -> Bool
isInvalid n = any isRepeated [1..len `div` 2]
  where
    s = show n
    len = length s
    isRepeated patLen = len `mod` patLen == 0 &&
                        all (== pattern) (chunks patLen s)
      where
        pattern = take patLen s

chunks :: Int -> [a] -> [[a]]
chunks _ [] = []
chunks n xs = take n xs : chunks n (drop n xs)
